File: packages/scout/src/cherrypick/scout/services/staging.py
"""Order dry-run validation and staged-ticket persistence -- the builder's "Validate with broker"
and "Stage" actions.

**This is the single broker-order call site in the whole package.** `dry_run_order` calls
`cherrypick.core.broker.place_order(..., live=False, ...)` with `live=False` a hardcoded literal --
never threaded through as a variable, a config value, or a request field. Every mutating caller
(`api/orders.py`'s dry-run route, and `stage_ticket` below) goes through this one function, so there
is exactly one place in this package that can ever reach the SDK's order-placement call, and it can
never submit live. `test_dry_run_only.py` both exercises this with a faked broker and source-scans
the package for that invariant.

Staging must not depend on validation succeeding (see the plan's risk #9): `stage_ticket` always
saves the ticket, even when `dry_run_order` itself fails (no credentials, a network hiccup, a
preflight rejection, an SDK response-shape drift) -- the failure is recorded as the ticket's
`dry_run` field rather than blocking the save. A research session shouldn't lose a ticket to a
broker hiccup on something the user just wants to park for later.
"""

from __future__ import annotations

import json
import sqlite3
import time
import uuid
from typing import Any

from cherrypick.core import broker as _broker

from .session import BrokerSession


def _mask_account(value: Any) -> str:
    s = str(value or "")
    return f"****{s[-4:]}" if len(s) >= 4 else "****"


def _serialize(obj: Any) -> Any:
    if obj is None or isinstance(obj, (str, int, float, bool)):
        return obj
    if isinstance(obj, (list, tuple)):
        return [_serialize(item) for item in obj]
    if isinstance(obj, dict):
        return {k: _serialize(v) for k, v in obj.items()}
    dump = getattr(obj, "model_dump", None)
    if callable(dump):
        return dump(mode="json")
    return str(obj)


def _leg_action(quantity: float) -> str:
    return "buy to open" if quantity > 0 else "sell to open"


def build_order_spec(legs: list[dict]) -> dict:
    """A leg basket (each ``{"symbol", "quantity", "price"}``; ``symbol`` is the OCC option symbol
    from `chain_service`, ``quantity`` signed -- positive long/buy, negative short/sell; ``price`` per
    share) -> `cherrypick.core.broker.build_order`'s dict spec. Every leg is opening (the builder
    stages new positions, never closes existing ones) and every leg is an option -- the builder has
    no stock-leg picker yet. Net price is per share, matching the SDK's own order-price convention --
    not the *100 dollar totals the screener/builder display panels show."""
    order_legs = [
        {
            "instrument_type": "Equity Option",
            "symbol": leg["symbol"],
            "action": _leg_action(leg["quantity"]),
            "quantity": abs(leg["quantity"]),
        }
        for leg in legs
    ]
    spec: dict = {"legs": order_legs}
    net = -sum(leg["quantity"] * leg["price"] for leg in legs)
    if net != 0:
        spec["price"] = round(abs(net), 2)
        spec["price_effect"] = "credit" if net >= 0 else "debit"
    return spec


async def dry_run_order(broker_session: BrokerSession, legs: list[dict]) -> dict:
    """Validate a leg basket against the broker's dry-run preflight -- buying-power effect, fees,
    warnings, **no order created**. Never raises: any failure (missing credentials, no accounts on
    the login, a network hiccup, an SDK response-shape change) comes back as
    ``{"ok": False, "error": ...}`` instead of propagating, so neither caller needs its own broad
    except around this. The account resolved is simply the first on the login (scout has no stored
    account-designation config of its own, unlike the trading modules) -- account numbers are masked
    before this ever returns."""
    if not legs:
        return {"ok": False, "error": "no legs to validate"}
    spec = build_order_spec(legs)

    async def _run(session: Any) -> dict:
        account = await _broker.resolve_account(session)
        order = _broker.build_order(spec)
        return await _broker.place_order(account, session, order, live=False, serialize=_serialize)

    try:
        result = await broker_session.call(_run)
    except Exception as exc:  # noqa: BLE001 -- soft-fail by design, see module/function docstrings
        return {"ok": False, "error": f"{type(exc).__name__}: {exc}"}

    if result.get("account_number") is not None:
        result["account_number"] = _mask_account(result["account_number"])
    return result


def _row_to_dict(row: tuple) -> dict:
    (id_, created_at, symbol, strategy, legs_json, credit, max_risk, dry_run_json, note, status) = row
    return {
        "id": id_,
        "created_at": created_at,
        "symbol": symbol,
        "strategy": strategy,
        "legs": json.loads(legs_json),
        "credit": credit,
        "max_risk": max_risk,
        "dry_run": json.loads(dry_run_json) if dry_run_json else None,
        "note": note,
        "status": status,
    }


def list_staged(conn: sqlite3.Connection) -> list[dict]:
    rows = conn.execute(
        "SELECT id, created_at, symbol, strategy, legs_json, credit, max_risk, dry_run_json, note, "
        "status FROM staged_orders ORDER BY created_at DESC"
    ).fetchall()
    return [_row_to_dict(r) for r in rows]


async def stage_ticket(
    conn: sqlite3.Connection,
    broker_session: BrokerSession,
    *,
    symbol: str,
    strategy: str,
    legs: list[dict],
    credit: float | None,
    max_risk: float | None,
    note: str | None,
    now: float | None = None,
) -> dict:
    """Validate (best-effort -- see `dry_run_order`) and persist a staged ticket. Always saves, even
    when the dry-run itself failed."""
    dry_run = await dry_run_order(broker_session, legs)
    ticket_id = uuid.uuid4().hex
    created_at = time.time() if now is None else now
    conn.execute(
        "INSERT INTO staged_orders (id, created_at, symbol, strategy, legs_json, credit, max_risk, "
        "dry_run_json, note, status) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, 'staged')",
        (
            ticket_id,
            created_at,
            symbol,
            strategy,
            json.dumps(legs),
            credit,
            max_risk,
            json.dumps(dry_run),
            note,
        ),
    )
    conn.commit()
    return _row_to_dict(
        (
            ticket_id,
            created_at,
            symbol,
            strategy,
            json.dumps(legs),
            credit,
            max_risk,
            json.dumps(dry_run),
            note,
            "staged",
        )
    )


def delete_staged(conn: sqlite3.Connection, ticket_id: str) -> bool:
    cur = conn.execute("DELETE FROM staged_orders WHERE id = ?", (ticket_id,))
    conn.commit()
    return cur.rowcount > 0

File: packages/console/server/src/services/staging.ts
/**
 * Order dry-run validation and staged-ticket persistence — the builder's
 * "Stage ticket" action.
 *
 * **This is the single broker-order call site in the whole package**, ported
 * from scout's staging.py with the same invariant: the ONLY order-shaped SDK
 * method this package ever calls is `postOrderDryRun` — buying-power effect,
 * fees, warnings, no order created. It is written here as a literal method
 * call, never threaded through as a variable or parameter, and a source-scan
 * test asserts no code path in this package can reach any order-creation or
 * order-mutation SDK method. Dry-run is still an authenticated
 * write-shaped call: button-triggered only, never on a timer or page load.
 *
 * Staging must not depend on validation succeeding: `stageTicket` always
 * saves, recording a dry-run failure (no credential, read-only-scope 403,
 * network hiccup) in the ticket's dryRun field instead of blocking the save.
 */

import { getClient, hasCredential } from "../market/session.js";
import { getScope } from "../auth/credentials.js";

export interface TicketLeg {
  /** OCC option symbol from the chain. */
  symbol: string;
  /** Signed: positive = buy to open, negative = sell to open. */
  quantity: number;
  /** Per share. */
  price: number;
  /** The structured fields Builder already has before it flattens a leg to an OCC symbol --
      carried through staging so a staged ticket can render expiry/strike/DTE the way Builder
      does, instead of asking a reader to decode the OCC symbol by hand. Optional and unused by
      the order spec/dry-run below; purely for staged-ticket display. */
  expiration?: string | null;
  strike?: number | null;
  kind?: "call" | "put" | "stock" | null;
}

function maskAccount(value: unknown): string {
  const s = String(value ?? "");
  return s.length >= 4 ? `****${s.slice(-4)}` : "****";
}

function maskAccounts(obj: unknown): unknown {
  if (Array.isArray(obj)) return obj.map(maskAccounts);
  if (typeof obj === "object" && obj !== null) {
    const out: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(obj)) {
      out[k] = k === "account-number" || k === "account_number" ? maskAccount(v) : maskAccounts(v);
    }
    return out;
  }
  return obj;
}

export function buildOrderSpec(legs: TicketLeg[]): Record<string, unknown> {
  const orderLegs = legs.map((leg) => ({
    "instrument-type": "Equity Option",
    symbol: leg.symbol,
    action: leg.quantity > 0 ? "Buy to Open" : "Sell to Open",
    quantity: Math.abs(leg.quantity),
  }));
  const spec: Record<string, unknown> = {
    "time-in-force": "Day",
    "order-type": "Limit",
    legs: orderLegs,
  };
  const net = -legs.reduce((s, leg) => s + leg.quantity * leg.price, 0);
  if (net !== 0) {
    spec["price"] = Math.round(Math.abs(net) * 100) / 100;
    spec["price-effect"] = net >= 0 ? "Credit" : "Debit";
  }
  return spec;
}

export interface DryRunResult {
  ok: boolean;
  error?: string;
  /** True when validation was skipped by policy (read-only credential), not by failure. */
  skipped?: boolean;
  account?: string;
  result?: unknown;
}

/** Never throws — every failure becomes {ok:false, error}. */
export async function dryRunOrder(legs: TicketLeg[]): Promise<DryRunResult> {
  if (legs.length === 0) return { ok: false, error: "no legs to validate" };
  if (!hasCredential()) return { ok: false, error: "no suite broker credential" };
  // Write-oriented functions disable themselves on a read-only credential:
  // the scope was detected by the live probe at boot / credentials set, so
  // don't spend a broker call on a request that will 403.
  if (getScope().scope === "read") {
    return { ok: false, skipped: true, error: "read-only credential — broker dry-run validation disabled" };
  }
  try {
    const client = getClient();
    const accounts = (await client.accountsAndCustomersService.getCustomerAccounts()) as Array<
      Record<string, unknown>
    >;
    const first = accounts?.[0]?.["account"] as Record<string, unknown> | undefined;
    const accountNumber = String(first?.["account-number"] ?? "");
    if (accountNumber === "") return { ok: false, error: "no accounts on this login" };
    const spec = buildOrderSpec(legs);
    // The single order-shaped broker call in this package. Dry-run only.
    const result: unknown = await client.orderService.postOrderDryRun(accountNumber, spec);
    return { ok: true, account: maskAccount(accountNumber), result: maskAccounts(result) };
  } catch (err) {
    const e = err as { response?: { status?: number; data?: unknown }; message?: string };
    const status = e.response?.status;
    const detail = e.response?.data !== undefined ? JSON.stringify(e.response.data).slice(0, 400) : e.message;
    return { ok: false, error: `${status !== undefined ? `HTTP ${status}: ` : ""}${detail ?? "unknown error"}` };
  }
}
